src/details/ArborX_DetailsDistributor.hpp

```diff
@@ -31,6 +31,94 @@ namespace ArborX
+  static_assert(
+      std::is_same<typename InputView::non_const_value_type, int>::value, "");
+  static_assert(std::is_same<typename OutputView::value_type, int>::value, "");
+
+  // In case all the batches are empty, return an empty list of unique_ranks and
+  // counts, but still have one element in offsets. This is conforming with
+  // creating the total offsets from batched_ranks and batched_offsets ignoring
+  // empty batches and calling sortAndDetermeineBufferLayout.
+  offsets.push_back(0);
+
+  auto const n_batched_ranks = batched_ranks.size();
+  if (n_batched_ranks == 0 || lastElement(batched_offsets) == 0)
+    return;
+
+  using DeviceType = typename InputView::traits::device_type;
+  using ExecutionSpace = typename InputView::traits::execution_space;
+
+  // Find the indices in batched_ranks for which the rank changes and store
+  // these ranks and the corresponding offsets in a new container that we can be
+  // sure to be large enough.
+  Kokkos::View<int *, DeviceType> compact_offsets(
+      Kokkos::ViewAllocateWithoutInitializing(batched_offsets.label()),
+      batched_offsets.size());
+  Kokkos::View<int *, DeviceType> compact_ranks(
+      Kokkos::ViewAllocateWithoutInitializing(batched_ranks.label()),
+      batched_ranks.size());
+
+  int n_unique_ranks;
+  Kokkos::parallel_scan(
+      ARBORX_MARK_REGION("compact_offsets_and_ranks"),
+      Kokkos::RangePolicy<ExecutionSpace>(0, n_batched_ranks),
+      KOKKOS_LAMBDA(unsigned int i, int &update, bool last_pass) {
+        if (i == batched_ranks.size() - 1 ||
+            batched_ranks(i + 1) != batched_ranks(i))
+        {
+          if (last_pass)
+          {
+            compact_ranks(update) = batched_ranks(i);
+            compact_offsets(update + 1) = batched_offsets(i + 1);
+          }
+          ++update;
+        }
+      },
+      n_unique_ranks);
+
+  // Now create subviews containing the elements we actually need, copy
+  // everything to the CPU and fill the output variables.
+  auto restricted_offsets = InputView(
+      compact_offsets, std::make_pair(1, static_cast<int>(n_unique_ranks + 1)));
+  auto restricted_unique_ranks = InputView(
+      compact_ranks, std::make_pair(0, static_cast<int>(n_unique_ranks)));
+
+  auto const unique_ranks_host = Kokkos::create_mirror_view_and_copy(
+      Kokkos::HostSpace(), restricted_unique_ranks);
+  unique_ranks.reserve(unique_ranks_host.size());
+  auto const offsets_host = Kokkos::create_mirror_view_and_copy(
+      Kokkos::HostSpace(), restricted_offsets);
+  offsets.reserve(restricted_offsets.size() + 1);
+  counts.reserve(restricted_offsets.size());
+
+  for (unsigned int i = 0; i < restricted_offsets.size(); ++i)
+  {
+    int const count =
+        i == 0 ? offsets_host(0) : offsets_host(i) - offsets_host(i - 1);
+    if (count > 0)
+    {
+      counts.push_back(count);
+      offsets.push_back(offsets_host(i));
+      unique_ranks.push_back(unique_ranks_host(i));
+    }
+  }
```

Comment:
I really don't understand what's going on in these lines. I also think there is some contradiction here: the comment says that we return offsets with at least one element, but that's not what's happening: if all counts are zero, nothing is pushed to offsets.

My original plan was to simplify the later part to
```c++
  // Copy data to CPU and fill out the output variables                                                                         
  auto const unique_ranks_host =                                                                                                
      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), compact_ranks);                                                  
  auto const offsets_host =                                                                                                     
      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), compact_offsets);                                                
                                                                                                                                
  unique_ranks.reserve(n_unique_ranks);                                                                                         
  offsets.reserve(n_unique_ranks + 1);                                                                                          
  counts.reserve(n_unique_ranks);                                                                                               
                                                                                                                                
  offsets.push_back(0);                                                                                                         
  for (unsigned int i = 0; i < n_unique_ranks; ++i)                                                                             
  {                                                                                                                             
    int const count = offsets_host(i + 1) - offsets_host(i);                                                                    
    if (count > 0)                                                                                                              
    {                                                                                                                           
      counts.push_back(count);                                                                                                  
      offsets.push_back(offsets_host(i + 1));                                                                                   
      unique_ranks.push_back(unique_ranks_host(i));                                                                             
    }                                                                                                                           
  }
```
but it hangs the test. Definitely need more understanding here.

Reply:
We push `0` to offsets near the beginning of the function (l.57). Apart from treating the case of empty ranges differently, the suggestion above misses treating the second offset correctly. Since `offsets_host` doesn't take into account the first element in `offsets` (which is 0), the first element in `counts` has to be `offsets_host(0)`. Also, `offsets_host` has size `n_unique_ranks`. Hence,  element `n_unique_ranks` is out-of-bounds. Basically, you everything needs to be shifted by one.

Reply:
OK. So the real problem is that the result of the parallel_scan still leaves compact_offsets(0) uninitialized, that breaks things. Thus, there is little point of changing code.